### ISSUE TITLE
Cosmos DB triggered function advances checkpoint after cancellation

### ISSUE THREAD

**Issue by MikeNicholls** [Azure/azure-webjobs-sdk-extensions#851]
Our Cosmos DB change feed triggered Function App is sometimes skipping processing of items in the source container. I believe this is due to the functions host being stopped during scaling and the change feed checkpoint still advancing.

#### Repro steps

1. Clone the sample Function App from https://github.com/MikeNicholls/change-feed-sample

2. Configure the Function App in `local.settings.json` and run it locally.

3. Create a document in Cosmos DB to trigger the function.

4. While the function is executing, initiate a graceful shutdown of the functions host, either by pressing Ctrl+C or raising a SIGINT (e.g. `killall -2 func`). Note the ID of the document that was being processed.

5. Run the Function App locally again.

#### Expected behavior

The function triggers again on the same Cosmos DB document.

#### Actual behavior

The function will not trigger again (or will move onto the next document, if there is one).

#### Known workarounds

By not passing a cancellation token into the function, it will try to continue on while the host is shutting down. Eventually it will either complete or be killed, and if killed will will eventually resume after restart once the lease expires.

I'm not sure how reliable this workaround is, I don't know if it's possible for the user function code to be terminated early while the checkpoint still advances during a forced shutdown.

#### Related information 

* Microsoft.Azure.WebJobs.Extensions.CosmosDB version 4.3.0
* Microsoft.Azure.Cosmos version 3.35.1
* Sample project: https://github.com/MikeNicholls/change-feed-sample

**Comment by ealsur** [comment]
It's unclear how this scenario skips items? The checkpoint of the progress happens after:

```
FunctionResult result = await this._executor.TryExecuteAsync(new TriggeredFunctionData() { TriggerValue = docs }, cancellationToken);
```

So that means the executor completed, right? At that point, why not checkpoint even if the process is shutting down? 

Unless the executor can return without the user code completing execution?

**Comment by MikeNicholls** [comment]
The user function code is being passed a cancellation token.  As the change feed items are delivered in batches it's not unusual for the user code to take some time to execute as it operates on each item, so checking the cancellation token seems like a responsible thing for it to do (e.g. to allow the function to be stopped gracefully instead of eventually timing out and being forcibly killed).

If the user code respects the provided token by cancelling any work in progress once cancellation is signalled, it will not be called again with the same items due to the checkpoint having been advanced after cancellation.  In short, the user code is being asked to please stop what it is doing, but if it does so it won't be given the opportunity to try again.

> Unless the executor can return without the user code completing execution?

That is probably a bigger and potentially more philosophical question, there was some discussion of this recently on https://github.com/Azure/azure-functions-host/issues/8994#issuecomment-1634850358.

Both the [Event Hubs](https://github.com/Azure/azure-sdk-for-net/pull/36432) and [Durable Task](https://github.com/Azure/azure-functions-durable-extension/blob/318384c04e27a6b1a8653115d09032c2155ba8b8/src/WebJobs.Extensions.DurableTask/Listener/FunctionExecutionHelper.cs#L38) extensions check for cancellation after calling `TryExecuteAsync`, so it seems like this may not be a unique issue.

**Comment by ealsur** [comment]
Thanks for the clarification. I took a look at the proposed PR, do you mind seeing if tests can be added to it to avoid future regressions?

**Comment by ealsur** [comment]
Proposed fix has been released in 4.3.1 version
